### PR TITLE
CAT: do not insert headers for gRPC requests

### LINKED ISSUE #1730: gRPC call causes "ArgumentError: 'X-NewRelic-ID' is an invalid header key, must match [a-z0-9-_.]+" with Google Ads API
## Description
When we call Google Ads API like that example, it causes `"ArgumentError: 'X-NewRelic-ID' is an invalid header key, must match [a-z0-9-_.]+"`.
https://github.com/googleads/google-ads-ruby/blob/d29519bc6c9ec23cad348b9b0ff5375c08a9fdf8/examples/account_management/list_accessible_customers.rb#L28-L40

## Expected Behavior (Acceptance Criteria)

no `ArgumentError` is thrown.

## [Troubleshooting](https://discuss.newrelic.com/t/troubleshooting-frameworks/108787) or [NR Diag](https://docs.newrelic.com/docs/using-new-relic/cross-product-functions/troubleshooting/new-relic-diagnostics) results

A stack trace from Rollbar is below:
```
File /usr/local/lib/ruby/gems/2.7.0/gems/grpc-1.50.0-x86_64-linux/src/ruby/lib/grpc/generic/active_call.rb:373 in run_batch
File /usr/local/lib/ruby/gems/2.7.0/gems/grpc-1.50.0-x86_64-linux/src/ruby/lib/grpc/generic/active_call.rb:373 in request_response
File /usr/local/lib/ruby/gems/2.7.0/gems/grpc-1.50.0-x86_64-linux/src/ruby/lib/grpc/generic/client_stub.rb:174 in block (2 levels) in request_response
File /usr/local/lib/ruby/gems/2.7.0/gems/grpc-1.50.0-x86_64-linux/src/ruby/lib/grpc/generic/interceptors.rb:181 in block in intercept!
File /usr/local/lib/ruby/gems/2.7.0/gems/google-ads-googleads-20.0.0/lib/google/ads/google_ads/interceptors/error_interceptor.rb:65 in request_response
File /usr/local/lib/ruby/gems/2.7.0/gems/grpc-1.50.0-x86_64-linux/src/ruby/lib/grpc/generic/interceptors.rb:175 in intercept!
File /usr/local/lib/ruby/gems/2.7.0/gems/grpc-1.50.0-x86_64-linux/src/ruby/lib/grpc/generic/client_stub.rb:173 in block in request_response
File /usr/local/lib/ruby/gems/2.7.0/gems/gapic-common-0.9.0/lib/gapic/grpc/service_stub/rpc_call.rb:121 in call
File /usr/local/lib/ruby/gems/2.7.0/gems/gapic-common-0.9.0/lib/gapic/grpc/service_stub.rb:156 in call_rpc
File /usr/local/lib/ruby/gems/2.7.0/gems/google-ads-googleads-20.0.0/lib/google/ads/google_ads/v12/services/customer_service/client.rb:322 in list_accessible_customers
File /usr/local/lib/ruby/gems/2.7.0/gems/google-ads-googleads-20.0.0/lib/google/ads/google_ads/service_wrapper.rb:53 in public_send
File /usr/local/lib/ruby/gems/2.7.0/gems/google-ads-googleads-20.0.0/lib/google/ads/google_ads/service_wrapper.rb:53 in method_missing
```

https://github.com/grpc/grpc/blob/91091e3668144de9c6aa392f496bb7639f7025a7/src/ruby/lib/grpc/generic/active_call.rb#L373

## Steps to Reproduce

Our logic is similar with the example code: https://github.com/googleads/google-ads-ruby/blob/d29519bc6c9ec23cad348b9b0ff5375c08a9fdf8/examples/account_management/list_accessible_customers.rb#L28-L40
We are also investigating reproduce condition.

## Your Environment

- Ruby 2.7.6
- newrelic_rpm (8.13.1)
- grpc (1.50.0)
- gapic-common (0.9.0)
- google-ads-googleads (20.0.0)

## Additional context
Add any other context about the problem here. For example, relevant community posts or support tickets.

- https://github.com/newrelic/newrelic-ruby-agent/issues/1337

Additional AC:
In the event that CAT is enabled, do not raise an error and handle the use case gracefully.

### PR DESCRIPTION
**Pull request by fallwith**
CAT: do not insert headers for outbound gRPC requests, as gRPC libraries will validate all headers against a regex pattern that the CAT headers do not currently satisfy (CAT uses capital letters).

resolves #1730

### CODE REVIEW

File: lib/new_relic/agent/distributed_tracing/cross_app_tracing.rb [hunk 1]
```diff
@@ -249,6 +251,10 @@ def assign_intrinsic_transaction_attributes(state)
           txn.attributes.add_intrinsic_attribute(:referring_transaction_guid, referring_guid)
         end
       end
+
+      def grpc_headers?(headers)
+        headers&.class&.name&.include?('::GRPC::')
```
Comment (kaylareopelle) [inline lib/new_relic/agent/distributed_tracing/cross_app_tracing.rb:256]: I love seeing these safe navigation operators!

File: test/multiverse/suites/agent_only/cross_application_tracing_test.rb [hunk 2]
```diff
@@ -61,6 +61,57 @@ def test_cross_app_error_attaches_process_id_to_intrinsics
     assert_includes attributes_for(last_traced_error, :intrinsic), :client_cross_process_id
   end
 
+  def test_grpc_headers_returns_nil_if_the_headers_object_is_nil
+    assert_nil tracer.send(:grpc_headers?, nil)
+  end
+
+  def test_grpc_headrs_returns_nil_if_the_headers_object_class_is_nil
```
Comment (kaylareopelle) [inline test/multiverse/suites/agent_only/cross_application_tracing_test.rb:68]:
```suggestion
  def test_grpc_headers_returns_nil_if_the_headers_object_class_is_nil
```

File: lib/new_relic/agent/distributed_tracing/cross_app_tracing.rb [hunk 3]
```diff
@@ -53,6 +52,9 @@ def cat_path_hash
 
       def insert_cross_app_header(headers)
         return unless CrossAppTracing.cross_app_enabled?
+        # do not insert CAT headers for gRPC requests
+        # https://github.com/newrelic/newrelic-ruby-agent/issues/1730
+        return if grpc_headers?(headers)
```
Comment (kaylareopelle) [inline lib/new_relic/agent/distributed_tracing/cross_app_tracing.rb:57]: It looks like "then" case coverage is missing for this line.

**Review by tannalynn — COMMENTED** [review]
Looks good! 
I am curious though why you went with this approach instead of adding a check in the grpc instrumentation to not call the insert headers method is cat was enabled?

**Comment by fallwith** [comment]
> Looks good! I am curious though why you went with this approach instead of adding a check in the grpc instrumentation to not call the insert headers method is cat was enabled?

@tannalynn looking at it again, I think we should have the check occur in the grpc instrumentation itself. Originally I wanted the scope to be as small as possible (just CAT headers) to preserve any potential good headers functionality, but now that I see that the scope is too small.

The gRPC instrumentation invokes `NewRelic::Agent::Transaction::ExternalRequestSegment#add_request_headers`.

That method does the following:

- Updates the segment name when a host can be obtained from the request
- Inserts the synthetics header (bad, won't satisfy gRPC header regex)
- Inserts the DT header (bad, won't satisfy gRPC header regex)

Currently I'm only addressing the 3rd item on the list. I need to get after the 2nd one as well. I could of course just prevent the call from being made entirely and skip all 3 items. What are your thoughts on that and the loss of item 1 for grpc?

File: lib/new_relic/agent/distributed_tracing/cross_app_tracing.rb (hunk 3 above)
Comment (fallwith) [reply lib/new_relic/agent/distributed_tracing/cross_app_tracing.rb:57]: Odd. There was an existing test for the presence of the headers that should continue to test the `return` not being reached, and there has been a new test added to test the `return` being reached. Both sides of the `if` conditional should be getting tested. I'll revisit after I move the check as Tanna is suggesting.

**Comment by tannalynn** [comment]
> > Looks good! I am curious though why you went with this approach instead of adding a check in the grpc instrumentation to not call the insert headers method is cat was enabled?
> 
> @tannalynn looking at it again, I think we should have the check occur in the grpc instrumentation itself. Originally I wanted the scope to be as small as possible (just CAT headers) to preserve any potential good headers functionality, but now that I see that the scope is too small.
> 
> The gRPC instrumentation invokes `NewRelic::Agent::Transaction::ExternalRequestSegment#add_request_headers`.
> 
> That method does the following:
> 
> * Updates the segment name when a host can be obtained from the request
> * Inserts the synthetics header (bad, won't satisfy gRPC header regex)
> * Inserts the DT header (bad, won't satisfy gRPC header regex)
> 
> Currently I'm only addressing the 3rd item on the list. I need to get after the 2nd one as well. I could of course just prevent the call from being made entirely and skip all 3 items. What are your thoughts on that and the loss of item 1 for grpc?

IMO I think that's fine. It should only be skipped if someone actually is using CAT, which is super old, deprecated, and not many people use it, so for the vast majority of people this change won't effect anything.

File: lib/new_relic/agent/distributed_tracing/cross_app_tracing.rb (hunk 1 above)
Comment (fallwith) [reply lib/new_relic/agent/distributed_tracing/cross_app_tracing.rb]: Removed with the [26e4415](https://github.com/newrelic/newrelic-ruby-agent/pull/1837/commits/26e4415437bb1e86f455b45bedc947507a7b0b93) rework.

File: lib/new_relic/agent/distributed_tracing/cross_app_tracing.rb (hunk 3 above)
Comment (fallwith) [reply lib/new_relic/agent/distributed_tracing/cross_app_tracing.rb]: Moot with the [26e4415](https://github.com/newrelic/newrelic-ruby-agent/pull/1837/commits/26e4415437bb1e86f455b45bedc947507a7b0b93) rework.

**Review by kaylareopelle — APPROVED** [review]
Nice refactor!

**Pull request merged**